go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,323 @@
+	// https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#:~:text=The%20subdirectory%20is%20named%20with%20the%20first%202%20characters%20of%20the%20SHA%2D1%2C%20and%20the%20filename%20is%20the%20remaining%2038%20characters.
+	var wg sync.WaitGroup
+	wg.Add(256)
+	for i := 0; i < 256; i++ {
+		prefixDir := filepath.Join(root, fmt.Sprintf("%02x", i))
+		go func() {
+			defer wg.Done()
+			_ = os.MkdirAll(prefixDir, os.ModePerm)
+			_ = filepath.WalkDir(prefixDir, func(path string, d fs.DirEntry, err error) error {
+				// We log and continue on all errors, because cache read errors are not critical.
+				if err != nil {
+					log.Errorf("Error reading cache directory: %v", err)
+					return nil
+				}
+				if d.IsDir() {
+					return nil
+				}
+				subdir := filepath.Base(filepath.Dir(path))
+				k, err := res.getKeyFromFileName(subdir + d.Name())
+				if err != nil {
+					log.Errorf("Error parsing cached file name %s: %v", path, err)
+					return nil
+				}
+				atime, err := GetLastAccessTime(path)
+				if err != nil {
+					log.Errorf("Error getting last accessed time of %s: %v", path, err)
+					return nil
+				}
+				it := &qitem{
+					key: k,
+					lat: atime,
+				}
+				size, err := res.getItemSize(k)
+				if err != nil {
+					log.Errorf("Error getting file size of %s: %v", path, err)
+					return nil
+				}
+				res.store.Store(k, it)
+				atomic.AddInt64(&res.sizeBytes, size)
+				res.mu.Lock()
+				heap.Push(res.queue, it)
+				res.mu.Unlock()
+				return nil
+			})
+		}()
+	}
+	wg.Wait()
+	go res.gc()
+	return res
+}
+
+func (d *DiskCache) getItemSize(k key) (int64, error) {
+	return k.digest.Size, nil
+}
+
+// Releases resources and terminates the GC daemon. Should be the last call to the DiskCache.
+func (d *DiskCache) Shutdown() {
+	d.shutdown <- true
+}
+
+func (d *DiskCache) TotalSizeBytes() uint64 {
+	return uint64(atomic.LoadInt64(&d.sizeBytes))
+}
+
+func (d *DiskCache) getKeyFromFileName(fname string) (key, error) {
+	pair := strings.Split(fname, ".")
+	if len(pair) != 2 {
+		return key{}, fmt.Errorf("expected file name in the form [ac_]hash/size, got %s", fname)
+	}
+	size, err := strconv.ParseInt(pair[1], 10, 64)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid size in digest %s: %s", fname, err)
+	}
+	dg, err := digest.New(pair[0], size)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid digest from file name %s: %v", fname, err)
+	}
+	return key{digest: dg}, nil
+}
+
+func (d *DiskCache) getPath(k key) string {
+	return filepath.Join(d.root, k.digest.Hash[:2], fmt.Sprintf("%s.%d", k.digest.Hash[2:], k.digest.Size))
+}
+
+func (d *DiskCache) StoreCas(dg digest.Digest, path string) error {
+	if dg.Size > int64(d.maxCapacityBytes) {
+		return fmt.Errorf("blob size %d exceeds DiskCache capacity %d", dg.Size, d.maxCapacityBytes)
+	}
+	it := &qitem{
+		key: key{digest: dg},
+		lat: time.Now(),
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	_, exists := d.store.LoadOrStore(it.key, it)
```

Comment:
d.store is already synchronized by it.mu. I'd vote to make store a normal map[digest]*qitem instead; it would be faster.